draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -717,6 +774,175 @@ Steps:
+covers the case of supersingular curves `E: y^2=x^3+B`.
+
+**Preconditions**
+
+This algorithm works for any Weierstrass curve over `F_q` such that `A=0` and
+`q=2 mod 3`.
+
+**Examples**
+
+- `y^2=x^3+1`
+
+**Algorithm**: map2curve_bf
+
+Input:
+
+ - `alpha`: an octet string to be hashed.
+ - `B`: the constant from the Weierstrass curve.
+
+Output:
+
+ - `(x,y)`: a point in E.
+
+Operations:
+
+~~~
+1. u = HashToBase(alpha)
+2. x = (u^2-B)^(1/3)
+3. Output (x,u)
+~~~
+
+**Implementation**
+
+The following procedure implements the Boneh-Franklin's algorithm in a
+straight-line fashion.
+
+~~~
+map2curve_bf(alpha)
+
+Input:
+
+ alpha: an octet string to be hashed.
+ B    : the constant from the Weierstrass curve.
+
+Output:
+
+ (x, y): a point in E
+
+Precomputations:
+
+1.  c = (2p-1)/3           // Integer arithmetic
+
+Steps:
+
+1.  u = HashToBase(alpha)  // {0,1}^* -> Fp
+2. t0 = u^2                // t0 = u^2
+3. t1 = t0-B               // t1 = u^2-B
+4.  x = t1^c               // x  = (u^2-B)^(1/3)
+5. Output (x,u)
+~~~
+
+
+### Fouque-Tibouchi Method {#ftpairing}
+
+The map2curve_ft(alpha) implements the Fouque-Tibouchi's method {{FT12}} which
+covers the case of pairing-friendly curves `E: y^2=x^3+B`.
+Note that for pairing curves the destination group is usually a subgroup of the
+curve, hence, a scalar multiplication by the cofactor will be required to send
+the point to the desired subgroup.
+
+**Preconditions**
+
+This algorithm works for any Weierstrass curve over `Fq` such that `A=0`, `1+B` is
+a non-zero square in the field, and `q=7 mod 12`. This covers the case
+`q=1 mod 3` not handled by Boneh-Franklin's method.
+
+**Examples**
+
+- SECP256K1 curve {{SEC2}}
+- BN curves {{BN05}}
+- KSS curves {{KSS08}}
+- BLS curves {{BLS01}}
+
+**Algorithm**: map2curve_ft
+
+Input:
+
+ - `alpha`: an octet string to be hashed.
+ - `B`: the constant from the Weierstrass curve.
+ - `s`: a constant equal to sqrt(-3) in the field.
+
+Output:
+
+ - (x,y): a point in E.
+
+Operations:
+
+~~~
+1. t = HashToBase(alpha)
+2. w = (st)/(1+B+t^2)
+3. x1 = (-1+s)/2-tw
+4. x2 = -1-x1
+5. x3 = 1+1/w^2
+6. e = Legendre(t)
+7. if x1^3+B is square, output (x1, e * sqrt(x1^3+B) )
+8. if x2^3+B is square, output (x2, e * sqrt(x2^3+B) )
+9. Output (x3, e * sqrt(x3^3+B))
+~~~
+
+**Implementation**
+
+The following procedure implements the Fouque-Tibouchi's algorithm in a
+straight-line fashion.
+
+~~~
+map2curve_ft(alpha)
+
+Input:
+
+  alpha: an octet string to be encoded
+  B    : the constant of the curve
+
+Output:
+
+  (x, y): - a point in E
+
+Precomputations:
+
+1.  c1 = sqrt(-3)           // Field arithmetic
+2.  c2 = (-1+c1)/2          // Field arithmetic
+
+Steps:
+
+1.   u = HashToBase(alpha)  // {0,1}^* -> Fp
+2.  t0 = u^2                // u^2
+3.  t0 = t0+B+1             // u^2+B+1
+4.  t0 = 1/t0               // 1/(u^2+B+1)
+5.  t0 = t0*u               // u/(u^2+B+1)
+6.  t0 = t0*c1              // sqrt(-3)u/(u^2+B+1)
```

Comment:
Can we use different values (and names -- not `ti`) for these intermediates? That would help with test vector generation.